### PR TITLE
add context manager functionality to ion and ioff

### PR DESCRIPTION
**Pull request by ianhi**
## PR Summary
Closes: https://github.com/matplotlib/matplotlib/issues/17013
Also closes: https://github.com/matplotlib/ipympl/issues/220

This allows the usage of `plt.ioff` and `plt.ion` as context managers:
```python
plt.ion()
with plt.ioff():
    # now interactive mode is off
    fig = plt.figure()
    ax = fig.gca()
```
```python
plt.ioff()
with plt.ion():
    # now interactive mode is on
    fig = plt.figure()
    ax = fig.gca()
```
It also caches the state and returns to the original state on exit:
```python
plt.ion()
with plt.ioff():
    # now interactive mode is off
    fig = plt.figure()
    ax = fig.gca()
assert plt.isinteractive
```
## Documentation - WIP -  advice welcome
I think that this additional behavior ought to be documented but I was a bit unsure how to go about it. In general things in the documentation seem divided into being functions or objects, and this PR puts `ion` and `ioff` somewhere between. So I think it makes sense to keep these on this page: https://matplotlib.org/3.2.1/api/_as_gen/matplotlib.pyplot.html?highlight=pyplot#module-matplotlib.pyplot. But, then they would be mislabelled as Functions, and I'm not sure that they will end up there anyway as that page is autogenerated. I also wasn't sure which docstring to put information about the context manager usage in, I think the most sensible place is in the `__call__` method, but this isn't strictly what is used as the context manager.

The documentation needs be updated at these links:  
https://matplotlib.org/3.2.1/users/shell.html#controlling-interactive-updating  
https://matplotlib.org/3.2.1/tutorials/introductory/usage.html?highlight=usage#what-is-interactive-mode

I'll give this a go but wanted to get input on where to put extended docstrings and how to manage the autogenerating docs. Also, if anyone has any wording suggestions I'd be happy to hear them.

## Tests
There don't seem to be any tests that make use of `ion` or `ioff`
```bash
cd lib/matplotlib/tests
grep -rni 'ioff()'
```
returns no results. Does this PR warrant adding new tests?

## PR Checklist

- [x] Has Pytest style unit tests 
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [x] New features are documented, with examples if plot related
    - See above discussion of documentation
- [x] Documentation is sphinx and numpydoc compliant
- [x] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] NA? Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by jklymak** [comment]
@tacaswell @QuLogic  @efiring   I don't understand `ion/ioff`, but maybe this PR could get some discussion?

**Comment by QuLogic** [comment]
Seems fine, but it's still draft...

**Comment by ianhi** [comment]
I rebased and started trying to build the docs in order to see how my change affected the docs. I followed the instructions here (https://matplotlib.org/3.1.1/devel/documenting_mpl.html) and even tried them in a fresh conda env:
```
sudo apt install graphviz texlive texlive-xetex
conda create -n mpl-dev python
pip install -e .
pip install -r requirements/doc/doc-requirements.txt
cd docs
make html
```

But I keep running into this mystifying error:
image

Any advice for fixing that error/discovering what the error is? It makes it much harder to work out the above issues with the docs. Though I can confirm from the [build artifact ](https://42480-1385122-gh.circle-artifacts.com/0/doc/build/html/search.html?q=ioff) that the ioff doc page no longer exists.

**Comment by ianhi** [comment]
@QuLogic sorry I didn't see your comment before I posted the above. The code changes work for daily usage (I have started using this by monkeypatching it in and using daily) but the remaining issue is that by making `ioff` and `ion` instances of a class it changes how they show up in the docs and I am not sure if these changes are ok, or if not how to fix them.

**Comment by ianhi** [comment]
I just rewrote this context manager in the same manner as the `xkcd` context manager which should return `ion` and `ioff` to being functions which will allow the docs to render correctly (still can't get them to build locally :disappointed: so pushing to see the docs build). I added tests and an api note + new docs and will update the original post's checklist accordingly. If all the tests pass then I think this will be ready for review

**Comment by ianhi** [comment]
@jklymak I think this is ready for review. Could you undo the marking as draft (I don't seem to be able to)

New documentation links:
[ioff](https://43065-1385122-gh.circle-artifacts.com/0/doc/build/html/api/_as_gen/matplotlib.pyplot.ioff.html?highlight=ioff#matplotlib.pyplot.ioff)
[ion](https://43065-1385122-gh.circle-artifacts.com/0/doc/build/html/api/_as_gen/matplotlib.pyplot.ion.html?highlight=ion#matplotlib.pyplot.ion)

**Comment by jklymak** [comment]
... The "Ready For Review" button doesn't work?  Happy to click it for you but just wanted to see what the interface is?

**Comment by ianhi** [comment]
> ... The "Ready For Review" button doesn't work? Happy to click it for you but just wanted to see what the interface is?

:facepalm: I was looking at the top right section of the page - totally the wrong place. Sorry about that and thanks for the tip

**Comment by jklymak** [comment]
NP, just didn't want to convert things to draft if the author couldn't un-draft it.....

File: doc/api/next_api_changes/behavior/17371-IHI.rst [hunk 1]
```diff
@@ -0,0 +1,7 @@
+ioff and ion can be used as context managers
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+``plt.ion()`` and ``plt.ioff()`` may now be used as context managers to create
+a context with interactive mode off, or on respectively. The old behavior of
```
Comment (QuLogic) [inline doc/api/next_api_changes/behavior/17371-IHI.rst:5]: The respectively clauses are reversed.
```suggestion
``plt.ion()`` and ``plt.ioff()`` may now be used as context managers to create
a context with interactive mode on or off, respectively. The old behavior of
```

File: lib/matplotlib/pyplot.py [hunk 2]
```diff
@@ -364,9 +364,41 @@ def isinteractive():
     return matplotlib.is_interactive()
 
 
+class _ioff():
```
Comment (QuLogic) [inline lib/matplotlib/pyplot.py:367]:
```suggestion
class _ioff:
```

File: lib/matplotlib/pyplot.py [hunk 3]
```diff
@@ -364,9 +364,41 @@ def isinteractive():
...
+        self.__call__()
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        if self.wasinteractive:
+            matplotlib.interactive(True)
+            install_repl_displayhook()
+        del self.wasinteractive
+
+
+class _ion():
```
Comment (QuLogic) [inline lib/matplotlib/pyplot.py:383]:
```suggestion
class _ion:
```

**Comment by ianhi** [comment]
> Does this work recursively? 

Yes it does. Would you like me to add that to tests?

It turns out that by using `call` rather than `init` I was breaking the normal usage of `ion` and `ioff` so I switched to using `init` and updated the tests so that they would catch that. One edge case here is: If you save the return value from a call to `ioff` or `ion` and try to use it as a contextmanager again it will no longer work. 
image

This is pretty niche and hopefully no one is saving the result of `plt.ioff` but perhaps I should add a `_used` variable that is set to true on `__exit__` and then raise a warning in `__enter__`? Like so:

```python
class _ioff:
    _used = False

    def __init__(self):
        self.wasinteractive = isinteractive()
        matplotlib.interactive(False)
        uninstall_repl_displayhook()

    def __enter__(self):
        if self._used:
            cbook._warn_external("re-using an ioff contextmanager is",
                                 " unsupported and may have unexpected results")

    def __exit__(self, exc_type, exc_value, traceback):
        if self.wasinteractive:
            matplotlib.interactive(True)
            install_repl_displayhook()
        self._used = True
```

**Comment by ianhi** [comment]
The tests all failed due to a primordial implementation of `_used` sneaking into the commit :/. I can fix that either by removing or implementing the code proposed in the above comment. So I'll wait for feedback re the warning

File: lib/matplotlib/pyplot.py [hunk 4]
```diff
@@ -375,25 +406,54 @@ def ioff():
...
+    be used as a context manager, for example::
+
+        with plt.ioff():
+            # interactive mode will be off
+            # figures will not automatically be shown
+            fig1 = plt.figure()
+            # ...
+
+        # This figure will be shown immediately
+        fig2 = plt.figure()
```
Comment (timhoffm) [inline lib/matplotlib/pyplot.py:422]: I would mention the return value in the notes section for the sake of being explicit.

~~~
    The return value is needed to make the function work as a context
    manager. It is not intended to be stored or accessed by the user.
~~~

File: lib/matplotlib/pyplot.py [hunk 5]
```diff
@@ -364,9 +364,40 @@ def isinteractive():
     return matplotlib.is_interactive()
 
 
+class _ioff:
+    def __init__(self):
```
Comment (timhoffm) [inline lib/matplotlib/pyplot.py:368]: Should get some explanatory docstring. Something along the lines of

~~~
    """
    Context manager for `.ioff`.
    
    The state is changed in ``__init__()`` instead of ``__enter__()``. The
    latter is a no-op. This allows using `.ioff` both as a function and
    as a context.
    """
~~~

**Comment by ianhi** [comment]
I added more information to docstrings. Thinking more about the previously proposed `_used` approach I realized it can't be used to protect against using the return value as a contextmanager. What would be necessary is to detect if the `__init__` method was called inside of a `with` context, but that doesn't seem possible. So this pathological case will remain possible:
image


An alternative to adding extra documentation and then hoping that users don't do this we could create a separate `plt.ioff_context` function. This would solve the above issue, but it would add cognitive load to the user. I at least find it more intuitive to call `with plt.ioff():` than `with plt.ioff_context():`. If the tradeoff of not having `ion/ioff` have return values is deemed worthwhile then I'm happy to switch to that approach.


As an aside: What is difference between single and double backticks as used in matplotlib docstrings? This [SO](https://stackoverflow.com/questions/22256995/restructuredtext-in-sphinx-and-docstrings-single-vs-double-back-quotes-or-back) question seems to imply that the meaning of single backticks can vary project to project.

**Comment by QuLogic** [comment]
This is reStructuredText; single are cross-references, double are code blocks.

**Comment by timhoffm** [comment]
For cross-references see also https://matplotlib.org/devdocs/devel/documenting_mpl.html#referring-to-other-code.

Let me think a little about returning a context manager vs. a separate function.

File: lib/matplotlib/pyplot.py [hunk 6]
```diff
@@ -364,9 +364,55 @@ def isinteractive():
     return matplotlib.is_interactive()
 
 
+class _ioff:
```
Comment (timhoffm) [inline lib/matplotlib/pyplot.py:367]:
```suggestion
class _IoffContext:
```

File: lib/matplotlib/pyplot.py [hunk 7]
```diff
@@ -375,25 +421,60 @@ def ioff():
...
+            # figures will not automatically be shown
+            fig1 = plt.figure()
+            # ...
+
+        # This figure will be shown immediately
+        fig2 = plt.figure()
+
+    The return value of `plt.ioff` is necessary to make the function work
+    as a context manager. It is not intended to be stored or accessed
+    by the user.
```
Comment (timhoffm) [inline lib/matplotlib/pyplot.py:440]:
```suggestion
    To enable usage as a context manager, this function returns a
    ``_IoffContext`` object. The return value is not intended to be stored
    or accessed by the user.
```

same for `ion()`.

File: lib/matplotlib/pyplot.py [hunk 8]
```diff
@@ -375,25 +421,60 @@ def ioff():
...
 
 def ion():
     """
-    Turn the interactive mode on.
+    Turn interactive mode on.
 
     See Also
     --------
-    ioff : disable interactive mode
+    ion : enable interactive mode
```
Comment (timhoffm) [inline lib/matplotlib/pyplot.py:451]: This change seems to be a copy-paste error.

File: lib/matplotlib/pyplot.py [hunk 9]
```diff
@@ -364,9 +364,55 @@ def isinteractive():
...
+    def __enter__(self):
+        pass
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        if self.wasinteractive:
+            matplotlib.interactive(True)
+            install_repl_displayhook()
+
+
+class _ion:
```
Comment (timhoffm) [inline lib/matplotlib/pyplot.py:390]:
```suggestion
class _IoffContext:
```

**Comment by ianhi** [comment]
> IMHO it is better to have a function with a (potentially unused) return value context object than creating additional functions ioff/ion_context().

@timhoffm my initial solution avoided the issue of the return value by making `ion/ioff` instances of a class that implemented `__call__`, `__enter__`, and `__exit__`. I've actually been using that myself in the meantime ([ref/code example](https://github.com/ianhi/mpl-interactions/blob/2210e1f67950b16f3c1cd1fb2e5724213c1f71d7/mpl_interactions/utils.py#L13-L28)) but for the life of my I couldn't figure out how to make it work with the documentation as this made `ion/ioff` no longer functions. My initial post on this PR has details on this. But I think that the solution we've ended up at should be just fine :). It is the same approach taken by the `xkcd` context manager and by `rc_context`. 


I've updated the documentation with your helpful clarifications, and I made the examples in `ion` and `ioff` have the same structure.

File: doc/api/next_api_changes/behavior/17371-IHI.rst [hunk 10]
```diff
@@ -0,0 +1,7 @@
+ioff and ion can be used as context managers
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+``plt.ion()`` and ``plt.ioff()`` may now be used as context managers to create
```
Comment (QuLogic) [inline doc/api/next_api_changes/behavior/17371-IHI.rst:4]: This should be cross-referencable now that you've made the change to fix the docs?
```suggestion
`plt.ion()` and `plt.ioff()` may now be used as context managers to create
```

File: doc/api/next_api_changes/behavior/17371-IHI.rst [hunk 11]
```diff
@@ -0,0 +1,7 @@
...
+a context with interactive mode on, or off respectively. The old behavior of
+calling these functions was maintained. To use the new functionality
```
Comment (QuLogic) [inline doc/api/next_api_changes/behavior/17371-IHI.rst:6]:
```suggestion
calling these functions is maintained. To use the new functionality
```

**Comment by ianhi** [comment]
I added variations on nested contexts all of which work. I did think of one more pathological case. What is the expected behavior for:
```python
plt.ioff()
with plt.ioff():
    plt.ion()
print(mpl.is_interactive())
```

currently the context manager will leave you in interactive mode at the end

image


It should be easy to change it so that interactive mode is off in the end, but I'm 50/50 on what's best.

**Comment by dopplershift** [comment]
I think the context manager should change state on entry into the block, and restore the previous state on exit. Therefore IMO at the end of your code interactive mode should be off.
